zmon_aws_agent/postgresql.py

```diff
@@ -58,3 +59,119 @@ def get_databases_from_clusters(pgclusters, infrastructure_account, region,
         logger.exception("Failed to make Database entities for PostgreSQL clusters!")
 
     return entities
+
+
+def collect_eip_addresses(infrastructure_account):
+    ec2 = boto3.client('ec2')
+
+    addresses = call_and_retry(ec2.describe_addresses)['Addresses']
+
+    return [a for a in addresses if a['NetworkInterfaceOwnerId'] == infrastructure_account.split(':')[1]]
+# FIXME: depend on region, too?
+
+
+def filter_asgs(infrastructure_account, asgs):
+    return [gr for gr in asgs
+            if gr['infrastructure_account'] == infrastructure_account
+            and 'spilo_cluster' in gr.keys()]
+
+
+def filter_instances(infrastructure_account, instances):
+    return [i for i in instances if i['infrastructure_account'] == infrastructure_account]
+
+
+def collect_launch_configurations(infrastructure_account):
+    asg = boto3.client('autoscaling')
+    lc_paginator = asg.get_paginator('describe_launch_configurations')
+    lcs = lc_paginator.paginate().build_full_result()['LaunchConfigurations']
+
+    user_data = {}
+
+    for lc in lcs:
+        # LaunchConfigurationName takes the form of spilo-your-cluster-AppServerInstanceProfile-66CCXX77EEPP
+        lc_name = '-'.join(lc['LaunchConfigurationName'].split('-')[1:-2])
+        user_data[lc_name] = lc['UserData']
+
+    return user_data
+
+
+def extract_eipalloc_from_lc(launch_configuration, cluster_name):
+    import yaml
+    import base64
+
+    lc = launch_configuration.get(cluster_name, '')
+
+    user_data = base64.decodebytes(lc.encode('utf-8')).decode('utf-8')
+    user_data = yaml.safe_load(user_data)
+
+    return user_data['environment'].get('EIP_ALLOCATION', '')
+
+
+def get_postgresql_clusters(region, infrastructure_account, asgs, insts):
+    entities = []
+
+    addresses = collect_eip_addresses(infrastructure_account)
+    spilo_asgs = filter_asgs(infrastructure_account, asgs)
+    instances = filter_instances(infrastructure_account, insts)
+    launch_configs = []
+
+    # we will use the ASGs as a skeleton for building the entities
+    for cluster in spilo_asgs:
+        cluster_name = cluster['spilo_cluster']
+
+        cluster_instances = []
+        eip = []
+        eip_allocation = []
+
+        for i in cluster['instances']:
+            instance_id = i['aws_id']
+
+            try:
+                i_data = [inst for inst in instances if inst['aws_id'] == instance_id][0]
+            except IndexError:
+                raise Exception(str(cluster_instances))
```

Comment:
We don't catch exceptions in ``main``, so this has the potential to bust the agent and disable syncing of other entities.

Reply:
I also suggest handling exceptions more throughly here, since we can get exceptions in helper functions (e.g. boto3 related, etc ...)

Reply:
Covering exceptions/negative cases in tests would help of course